tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type.bal

```diff
@@ -191,3 +191,40 @@ public function testReadOnlyIntersectionFieldInRecord() {
     Z z = {y};
     assertEquality(z.y, y);
 }
+
+type Atom readonly & object {
+    function compare(Atom other);
+};
+
+readonly class AtomImpl {
+    *Atom;
+
+    function compare(Atom other) {
+        if (self === other) {
+            panic error("Same");
+        }
+    }
+}
+
+function testRecursiveReadonlyIntersection() {
+    Atom a = new AtomImpl();
+    error? e = trap a.compare(a);
+    if (e is error && e.message() == "Same") {
+        return;
+    }
+    panic error("Assertion error");
+}
+
+function testRuntimeTypeNameOfIntersectionType() {
+    any a = new AtomImpl();
+
+    error|int r = trap (<int> a);
+    if r is error {
+        var message = r.detail()["message"];
+        if message is string && message == "incompatible types: 'AtomImpl' cannot be cast to 'int'" {
+            return;
+        }
+    }
+    panic error("Assertion error");
```

Comment:
Shall we add more info re: why the test failed? If it was because the value was not an error vs message not being a string vs a message mismatch?

Since we can work with the assumption that `r` would definitely be an error with a string message here, IMO, we should consider changing L221 to

```ballerina
error r = <error> a;
```

and then maybe directly assert 
```ballerina
string message = <string> checkpanic r.detail()["message"];

if  message == "incompatible types: 'AtomImpl' cannot be cast to 'int'" {
   return;
}

panic error(string `expected 'incompatible types: 'AtomImpl' cannot be cast to 'int'', found '${message}'`);
```

IMO, then it will be clearer at first glance what the test is actually asserting.